optimize: Optimize RM/TM client reconnect logic

lokidundun opened the pull request:
<!--
    Licensed to the Apache Software Foundation (ASF) under one or more
    contributor license agreements.  See the NOTICE file distributed with
    this work for additional information regarding copyright ownership.
    The ASF licenses this file to You under the Apache License, Version 2.0
    (the "License"); you may not use this file except in compliance with
    the License.  You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0
    
    Unless required by applicable law or agreed to in writing, software
    distributed under the License is distributed on an "AS IS" BASIS,
    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
    See the License for the specific language governing permissions and
    limitations under the License.
-->
<!-- Please make sure you have read and understood the contributing guidelines -->

- [x] I have read the [CONTRIBUTING.md](https://github.com/apache/incubator-seata/blob/2.x/CONTRIBUTING.md) guidelines.
- [x] I have registered the PR [changes](https://github.com/apache/incubator-seata/tree/2.x/changes).

### Ⅰ. Describe what this PR did
- Optimize RM/TM client reconnect logic
- Replaces individual reconnect timers with a shared global timer for all RM/TM clients
- Maintains separate channel management for RM and TM while optimizing timer resources

### Ⅱ. Does this pull request fix one issue?
<!-- If that, add "fixes #xxx" below in the next line, for example, fixes #97. -->
fixes https://github.com/apache/incubator-seata/issues/5338

### Ⅲ. Why don't you add test cases (unit test/integration test)? 


### Ⅳ. Describe how to verify it


### Ⅴ. Special notes for reviews

lokidundun committed:
optimize: Optimize RM/TM client reconnect logic

lokidundun committed:
log change

YvCeung commented on core/src/main/java/org/apache/seata/core/rpc/netty/AbstractNettyRemotingClient.java:
```diff
@@ -155,6 +157,28 @@ public AbstractNettyRemotingClient(
+        Configuration configuration = ConfigurationFactory.getInstance();
+        this.enableReconnect =
+                configuration.getBoolean("client.reconnect.enable", DefaultValues.DEFAULT_ENABLE_CLIENT_RECONNECT);
+
```
Although there is currently a passive reconnection mechanism (on-demand reconnection), it inevitably increases request latency, as the connection is only detected and re-established when a request is actually sent.  I would like to understand: if the scheduled reconnection switch is disabled, are there any other negative impacts beyond the increased request latency?  Additionally, is it necessary to give users the decision-making authority over this switch?

lokidundun replied:
I haven't fully figured this out yet. Let me think about it more. Thank you for your suggestion, and I will value it seriously.

lokidundun replied:
I agree that exposing this switch to users might introduce unnecessary complexity，I will remove it,Thank you for your insightful feedback—it really helped refine the design.

lokidundun committed:
use global timer for RM/TM client reconnect to reduce resource usage

lokidundun committed:
Merge branch '2.x' into feature/client-reconnect-optimize

lokidundun committed:
remove the import*

lokidundun committed:
Merge branch 'feature/client-reconnect-optimize' of https://github.com/lokidundun/incubator-seata into feature/client-reconnect-optimize

lokidundun committed:
fix test fail

lokidundun committed:
run mvn spotless

lokidundun committed:
Merge branch '2.x' into feature/client-reconnect-optimize

lokidundun commented:
@YvCeung  PTAL❤️

lokidundun committed:
Merge branch '2.x' into feature/client-reconnect-optimize

lokidundun committed:
Merge branch '2.x' into feature/client-reconnect-optimize

maple525866 committed:
Merge branch '2.x' into feature/client-reconnect-optimize

lokidundun committed:
Merge branch '2.x' into feature/client-reconnect-optimize

maple525866 commented on core/src/main/java/org/apache/seata/core/rpc/netty/AbstractNettyRemotingClient.java:
Is there a risk of resource leaks here? Currently, `destroy()` only removes the instance from `CLIENT_INSTANCES`, but does not stop `GLOBAL_RECONNECT_TIMER` when there are no instances. Even if all clients are destroyed, the scheduler remains alive and occupies threads.

maple525866 commented:
At the same time, you should also increase testing to verify the correctness of the changes. @lokidundun

lokidundun committed:
optimize the destroy logic

lokidundun committed:
Merge branch 'feature/client-reconnect-optimize' of https://github.com/lokidundun/incubator-seata into feature/client-reconnect-optimize

lokidundun commented:
> At the same time, you should also increase testing to verify the correctness of the changes. @lokidundun

Thanks for the hint! I've added shutdownGlobalTimerIfNoClients() in destroy(), wired up the global timer/worker lifecycle and added tests covering the path. I'll further harden shutdown/wait logic and concurrency handling — please take another look when you have time.

lokidundun committed:
fix ci

lokidundun committed:
fix ci

lokidundun committed:
check pmd

lokidundun committed:
test:add some tests

maple525866 commented on core/src/main/java/org/apache/seata/core/rpc/netty/AbstractNettyRemotingClient.java:
```diff
@@ -773,4 +815,109 @@ public void close(ChannelHandlerContext ctx, ChannelPromise future) throws Excep
+                } catch (Exception ex) {
+                    LOGGER.warn("Shutdown global reconnect timer failed: {}", ex.getMessage(), ex);
+                }
+            }
```
Currently, `shutdownGlobalTimerIfNoClients()` appears to have a race condition. There's a window of opportunity: if another thread immediately creates a client and calls `add(this)` after `isEmpty()` returns true, the timer and worker will be stopped, but there will still be an active client. As a result, the new client won't be reconnected by the timer.

lokidundun replied:
Thank you for identifying the race condition. I have tried to resolve it by introducing a global lock and a double-check mechanism to ensure atomicity when checking client instances and shutting down resources.
Key Fixes:
- Added private static final ReentrantLock GLOBAL_SHUTDOWN_LOCK to synchronize the entire operation of checking CLIENT_INSTANCES.isEmpty() and shutting down the timer/worker.

Please take another look when you have time❤️.

maple525866 commented on core/src/main/java/org/apache/seata/core/rpc/netty/AbstractNettyRemotingClient.java:
```diff
@@ -773,4 +815,109 @@ public void close(ChannelHandlerContext ctx, ChannelPromise future) throws Excep
+            if (worker != null) {
+                try {
+                    worker.shutdown();
+                    worker.awaitTermination(1, TimeUnit.SECONDS);
```
Logically speaking, the `worker` also has this problem.

lokidundun committed:
fix race condition

lokidundun committed:
move the wrong tests

lokidundun committed:
fix

lokidundun committed:
run mvn

lokidundun commented:
maybe this pr is turning into comlexity,I will close it and open a new one
